src/platform/ASR/PlatformManagerImpl.cpp

```diff
@@ -0,0 +1,353 @@
+    {
+        ChipLogError(DeviceLayer, "lega_rtos_push_to_queue %u", result);
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    PlatformMgrImpl().SetEventFlags(kPostEventFlag);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR PlatformManagerImpl::_StartChipTimer(System::Clock::Timeout durationMS)
+{
+    if (durationMS.count() == 0)
+    {
+        TimerCallback(0);
+    }
+    else
+    {
+        lega_rtos_deinit_timer(&mTimer);
+        lega_rtos_init_timer(&mTimer, durationMS.count(), (timer_handler_t) TimerCallback, (void *) this);
+        OSStatus result = lega_rtos_start_timer(&mTimer);
+        if (kNoErr != result)
+        {
+            ChipLogError(DeviceLayer, "wiced_start_timer 0x%02x", result);
+            return CHIP_ERROR_INTERNAL;
+        }
+    }
+    return CHIP_NO_ERROR;
+}
+#else
+CHIP_ERROR PlatformManagerImpl::InitLwIPCoreLock(void)
+{
+    return Internal::InitLwIPCoreLock();
+}
+#endif
+
+void PlatformManagerImpl::_Shutdown()
+{
+    uint64_t upTime = 0;
+
+    if (GetDiagnosticDataProvider().GetUpTime(upTime) == CHIP_NO_ERROR)
+    {
+        uint32_t totalOperationalHours = 0;
+
+        if (ConfigurationMgr().GetTotalOperationalHours(totalOperationalHours) == CHIP_NO_ERROR)
+        {
+            ConfigurationMgr().StoreTotalOperationalHours(totalOperationalHours + static_cast<uint32_t>(upTime / 3600));
+        }
+        else
+        {
+            ChipLogError(DeviceLayer, "Failed to get total operational hours of the Node");
+        }
+    }
+    else
+    {
+        ChipLogError(DeviceLayer, "Failed to get current uptime since the Node’s last reboot");
+    }
+#if CONFIG_ENABLE_ASR_LEGA_RTOS
+    //
+    // Call up to the base class _Shutdown() to perform the actual stack de-initialization
+    // and clean-up
+    //
+
+    (void) _StopEventLoopTask();
+
+    // the task thread is self terminating, we might have to wait if it's still processing
+    while (true)
+    {
+        if (kTaskRunningEventFlag == 0)
+        {
+            break;
+        }
+        lega_rtos_delay_milliseconds(1);
+    }
+
+    ChipLogError(DeviceLayer, "StopEventLoopTask done.");
+
+    lega_rtos_deinit_event_flags(&mEventFlags);
+    lega_rtos_deinit_queue(&mEventQueue);
+    lega_rtos_deinit_timer(&mTimer);
+    lega_rtos_deinit_mutex(&mChipMutex);
+    lega_rtos_deinit_mutex(&mEventMutex);
+
+    Internal::GenericPlatformManagerImpl<PlatformManagerImpl>::_Shutdown();
+#else
+    Internal::GenericPlatformManagerImpl_FreeRTOS<PlatformManagerImpl>::_Shutdown();
+#endif
+}
+
+#if CONFIG_ENABLE_ASR_LEGA_RTOS
+void PlatformManagerImpl::SetEventFlags(uint32_t flags)
+{
+    if (lega_rtos_set_event_flags(&mEventFlags, flags) != kNoErr)
+    {
+        ChipLogError(DeviceLayer, "%s lega_rtos_set_event_flags %08lx", __func__, flags);
+    }
+}
+
+void PlatformManagerImpl::HandleTimerEvent(void)
+{
+    const CHIP_ERROR err = static_cast<System::LayerImplFreeRTOS &>(DeviceLayer::SystemLayer()).HandlePlatformTimer();
```

Comment:
Should mChipMutex get locked somewhere here?  @tx2rx @andy31415

Reply:
@bzbarsky-apple Same reason as above. if there is a problem with this design, let us will discuss it again

Reply:
@tx2rx Can you please describe how synchronization of the application with the Matter thread is supposed to be accomplished?

Concretely, if an application wants to write an attribute value, what it is expected to do on this platform?

Reply:
Or put another way, why are LockChipStack/UnlockChipStack as implemented here useful?

Reply:
yes, Lock is a good solution to the problem of synchronizing multiple threads , LockChipStack/UnlockChipStack haven't used it before.  resolved at https://github.com/project-chip/connectedhomeip/pull/26645